ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpRequestBodyPublisher.kt

```diff
@@ -0,0 +1,175 @@
+    ) : Flow.Subscription, CoroutineScope {
+        private val outstandingDemand = atomic(0L)
+        private var writeInProgress = false
+
+        @Volatile
+        private var done = false
+
+        override fun request(n: Long) {
+            if (done) {
+                return
+            }
+            if (n < 1) {
+                val ex = IllegalArgumentException(
+                    "$subscriber violated the Reactive Streams rule 3.9 by requesting "
+                        + "a non-positive number of elements."
+                )
+                signalOnError(ex)
+            } else {
+                try {
+                    // As governed by rule 3.17, when demand overflows `Long.MAX_VALUE` we treat the signalled demand as
+                    // "effectively unbounded"
+                    outstandingDemand.getAndUpdate { initialDemand: Long ->
+                        if (Long.MAX_VALUE - initialDemand < n) {
+                            Long.MAX_VALUE
+                        } else {
+                            initialDemand + n
+                        }
+                    }
+                    synchronized(this) {
+                        if (!writeInProgress) {
+                            writeInProgress = true
+                            readData()
+                        }
+                    }
+                } catch (e: Exception) {
+                    signalOnError(e)
+                }
+            }
+        }
+
+        override fun cancel() {
+            synchronized(this) {
+                if (!done) {
+                    done = true
+                    closeChannel()
+                }
+            }
+        }
+
+        private fun readData() {
+            // It's possible to have another request for data come in after we've closed the file.
+            if (inputChannel.isClosedForRead) {
+                return
+            }
+
+            launch {
+                do {
+                    val buffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE)
+                    val result = try {
+                        inputChannel.readAvailable(buffer)
+                    } catch (e: Throwable) {
+                        signalOnError(e)
+                        closeChannel()
+                        return@launch
+                    }
+
+                    if (result > 0) {
+                        buffer.flip()
+                        signalOnNext(buffer)
+                    }
+                    // If we have more permits, queue up another read.
+                } while (outstandingDemand.decrementAndGet() > 0)
+
+                if (inputChannel.isClosedForRead) {
+                    // Reached the end of the channel, notify the subscriber and cleanup
+                    signalOnComplete()
+                    closeChannel()
+                }
+
+                synchronized(this) {
+                    writeInProgress = false
```

Comment:
There is a race here: when we request, we increment the number of permits first (`outstandingDemand`) and then check for the flag. Here we do the opposite. Consider the case: we are already saw 0 count here and going to make `writeInProgress = false` but it is `true` yet and we are just before the synchronized block and lost the time quant. The "requester" comes just after and increases the count and looking at the flag and it's true so no new coroutine is launched. After that, we are putting false to `writeInProgress` and them we quit. So we have a positive `outstandingDemand` but no coroutine is handling it. The quick solution is to re-check the count before existing the launch but avoiding double-launch.

Reply:
Thanks for noticing this race condition. I've switched to using optimistic locking to avoid that:
https://github.com/ktorio/ktor/blob/4f1015e3a0ca4f7ac5dcf0725d2e3a7cfa4f1313/ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpRequestBodyPublisher.kt#L115